libbeat/processors/fingerprint/fingerprint.go

```diff
@@ -0,0 +1,134 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package fingerprint
+
+import (
+	"fmt"
+	"hash"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/processors"
+	jsprocessor "github.com/elastic/beats/libbeat/processors/script/javascript/module/processor"
+)
+
+func init() {
+	processors.RegisterPlugin("fingerprint", New)
+	jsprocessor.RegisterPlugin("Fingerprint", New)
+}
+
+const processorName = "fingerprint"
+
+var errNoFields = errors.New("must specify at least one field")
+
+type fingerprint struct {
+	config Config
+	fields []string
+	hash   hash.Hash
+}
+
+// New constructs a new fingerprint processor.
+func New(cfg *common.Config) (processors.Processor, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack %v processor configuration", processorName)
+	}
+
+	sort.Strings(config.Fields)
+
+	p := &fingerprint{
+		config: config,
+		hash:   config.Method(),
+		fields: unique(config.Fields),
+	}
+
+	return p, nil
+}
+
+// Run enriches the given event with fingerprint information
+func (p *fingerprint) Run(event *beat.Event) (*beat.Event, error) {
+	source, err := makeSource(p.fields, event.Fields)
+	if err != nil {
+		return nil, makeComputeFingerprintError(err)
+	}
+
+	hashFn := p.hash
+	hashFn.Reset()
+
+	hashFn.Write([]byte(source))
+	hash := hashFn.Sum(nil)
+
+	encodedHash := p.config.Encoding(hash)
+
+	if _, err = event.PutValue(p.config.TargetField, encodedHash); err != nil {
+		return nil, makeComputeFingerprintError(err)
+	}
+
+	return event, nil
+}
+
+func (p *fingerprint) String() string {
+	return fmt.Sprintf("%v=[method=[%v]]", processorName, p.config.Method)
+}
+
+func makeSource(sourceFields []string, eventFields common.MapStr) (string, error) {
+	var str string
+	for _, k := range sourceFields {
+		v, err := eventFields.GetValue(k)
+		if err == common.ErrKeyNotFound {
+			return "", errors.Wrapf(err, "failed to find field [%v] in event", k)
+		}
+		if err != nil {
+			return "", errors.Wrapf(err, "failed when finding field [%v] in event", k)
+		}
```

Comment:
Do we want to have an option to ignore missing fields in case we have at least one field present?

Reply:
Sorry, but is this the same as your suggestion in https://github.com/elastic/beats/pull/14205#discussion_r340863983 or something different?

Reply:
almost. My suggestion originally was to add support to ignore missing fields. But apparently we can have other error types as well. Would it make sense to treat those other types as 'missing' as well?

Reply:
I see. So if we can't "get" a field for whatever reason, we treat it as missing and then, if the `missing_fields` option is set, ignore it. Hmm, I think this makes sense but let me just look into what other types of errors (besides `common.ErrKeyNotFound`) might be returned here.

Reply:
The only other error that can be returned here is if we try to get a value for a nested field (e.g. `a.b.c`), but the ancestor path to the field (`a.b`) does not resolve to a `map`. To me this also feels like a missing field as, even in this case, we still could not find `a.b.c` for the user. So I'm okay with collapsing the two error cases into one and adding `ignore_missing` handling to it.